flink/src/main/java/org/apache/iceberg/flink/data/ParquetWithFlinkSchemaVisitor.java

```diff
@@ -19,181 +19,48 @@
-      }
-      Preconditions.checkArgument(sType instanceof RowType, "Invalid struct: %s is not a struct", sType);
-      RowType struct = (RowType) sType;
-      return visitor.struct(struct, group, visitFields(struct, group, visitor));
+  @Override
+  protected LogicalType arrayElementType(LogicalType arrayType) {
+    if (arrayType == null) {
+      return null;
     }
-  }
 
-  private static <T> T visitField(RowType.RowField sField, Type field, ParquetWithFlinkSchemaVisitor<T> visitor) {
-    visitor.fieldNames.push(field.getName());
-    try {
-      return visit(sField.getType(), field, visitor);
-    } finally {
-      visitor.fieldNames.pop();
-    }
+    return ((ArrayType) arrayType).getElementType();
   }
 
-  private static <T> List<T> visitFields(RowType struct, GroupType group,
-                                         ParquetWithFlinkSchemaVisitor<T> visitor) {
-    List<RowType.RowField> sFields = struct.getFields();
-    Preconditions.checkArgument(sFields.size() == group.getFieldCount(),
-        "Structs do not match: %s and %s", struct, group);
-    List<T> results = Lists.newArrayListWithExpectedSize(group.getFieldCount());
-    for (int i = 0; i < sFields.size(); i += 1) {
-      Type field = group.getFields().get(i);
-      RowType.RowField sField = sFields.get(i);
-      Preconditions.checkArgument(field.getName().equals(AvroSchemaUtil.makeCompatibleName(sField.getName())),
-          "Structs do not match: field %s != %s", field.getName(), sField.getName());
-      results.add(visitField(sField, field, visitor));
+  @Override
+  protected LogicalType mapKeyType(LogicalType mapType) {
+    if (mapType == null) {
+      return null;
     }
 
-    return results;
-  }
-
-  public T message(RowType sStruct, MessageType message, List<T> fields) {
-    return null;
-  }
-
-  public T struct(RowType sStruct, GroupType struct, List<T> fields) {
-    return null;
+    return ((MapType) mapType).getKeyType();
   }
 
-  public T list(ArrayType sArray, GroupType array, T element) {
-    return null;
-  }
-
-  public T map(MapType sMap, GroupType map, T key, T value) {
-    return null;
-  }
+  @Override
+  protected LogicalType mapValueType(LogicalType mapType) {
+    if (mapType == null) {
+      return null;
+    }
 
-  public T primitive(LogicalType sPrimitive, PrimitiveType primitive) {
-    return null;
+    return ((MapType) mapType).getValueType();
   }
 
-  protected String[] currentPath() {
-    return Lists.newArrayList(fieldNames.descendingIterator()).toArray(new String[0]);
-  }
+  @Override
+  protected LogicalType fieldType(LogicalType structType, int pos, Integer fieldId) {
+    if (structType == null || ((RowType) structType).getFieldCount() < pos + 1) {
+      return null;
+    }
 
-  protected String[] path(String name) {
-    List<String> list = Lists.newArrayList(fieldNames.descendingIterator());
-    list.add(name);
-    return list.toArray(new String[0]);
+    return ((RowType) structType).getTypeAt(pos);
```

Comment:
I'm also curious whether the inner fields are keep the same order as the parquet's fields.  If not, then we would get the incorrect data type by the `pos` provided from parquet.